lib/ohai/mixin/azure_metadata.rb

```diff
@@ -19,30 +19,87 @@
+                                   2018-10-01 2019-02-01 2019-03-11 2019-04-30 2019-06-01
+                                   2019-06-04 2019-08-01 2019-08-15 2019-11-01 }.freeze
+
+      def best_api_version
+        @api_version ||= begin
+          logger.trace("Mixin AzureMetadata: Fetching http://#{AZURE_METADATA_ADDR}/metadata/instance to determine the latest supported metadata release")
+          response = http_get("/metadata/instance")
+          if response.code == "404"
+            logger.trace("Mixin AzureMetadata: Received HTTP 404 from metadata server while determining API version, assuming #{AZURE_SUPPORTED_VERSIONS.last}")
+            return AZURE_SUPPORTED_VERSIONS.last
+          elsif response.code != "400" # 400 is actually what we want
+            raise "Mixin AzureMetadata: Unable to determine Azure metadata version (returned #{response.code} response)"
+          end
+
+          # azure returns a list of the 3 latest versions it supports
+          versions = parse_json(response.body).fetch("newest-versions", [])
+          versions.sort!
+
+          until versions.empty? || AZURE_SUPPORTED_VERSIONS.include?(versions.last)
+            pv = versions.pop
+            logger.trace("Mixin AzureMetadata: Azure metadata version #{pv} is not present in the versions provided by the Azure Instance Metadata service")
+          end
+
+          if versions.empty?
+            logger.debug "Mixin AzureMetadata: The short list of supported versions provided by Azure Instance Metadata service doesn't match any known versions to Ohai. Using the latest supported release known to Ohai instead: #{AZURE_SUPPORTED_VERSIONS.last}"
+            return AZURE_SUPPORTED_VERSIONS.last
+          end
+
+          logger.trace("Mixin AzureMetadata: Latest supported Azure metadata version: #{versions.last}")
+          versions.last
+        end
+      end
+
+      # fetch the meta content with a timeout and the required header and a read timeout of 6s
+      #
+      # @param [String] the relative uri to fetch from the Azure Metadata Service URL
+      #
+      # @return [Net::HTTP]
       def http_get(uri)
+        full_uri = URI.join("http://#{AZURE_METADATA_ADDR}", uri)
         conn = Net::HTTP.start(AZURE_METADATA_ADDR)
         conn.read_timeout = 6
-        conn.get(uri, { "Metadata" => "true" })
+        conn.get(full_uri, { "Metadata" => "true" })
       end
 
-      def fetch_metadata
-        logger.trace("Mixin AzureMetadata: Fetching metadata from host #{AZURE_METADATA_ADDR} at #{AZURE_METADATA_URL}")
-        response = http_get(AZURE_METADATA_URL)
+      # parse JSON data from a String to a Hash
+      #
+      # @param [String] response_body json as string to parse
+      #
+      # @return [Hash]
+      def parse_json(response_body)
+        data = StringIO.new(response_body)
+        parser = FFI_Yajl::Parser.new
+        parser.parse(data)
```

Comment:
FWIW, a string or StringIO works here.

Reply:
That's probably just cargo culting from ec2 metadata

Reply:
I updated all 3 of the metadata mixins to use just String